### ISSUE TITLE
Kong requeuing sync status, err Unauthorized

### ISSUE THREAD

**Issue by yousafsyed-bg** [Kong/kubernetes-ingress-controller#525]
### Summary
We have kong `1.3` running in our production with kong-ingress-conrtoller `0.6.0` with a `dbless` mode. It was working fine from last 3-4 months until it failed today with the following errors.

### Errors from ingress controller
```json
{
   "insertId": "umlngu6lf9ji0l946",
   "jsonPayload": {
     "source": "queue.go:113",
     "message": "requeuing sync status, err Unauthorized",
     "pid": "1"
   },
   "resource": {
     "type": "k8s_container",
     "labels": {
       "project_id": "brokergenius-qa-env",
       "cluster_name": "k8s-qa",
       "pod_name": "gateway-kong-5b757c4bb8-hx44g",
       "container_name": "ingress-controller",
       "namespace_name": "gateway",
       "location": "us-east1-b"
     }
   },
   "timestamp": "2020-02-03T09:48:09.073539Z",
   "severity": "WARNING",
   "labels": {
     "k8s-pod/component": "app",
     "k8s-pod/release": "gateway",
     "k8s-pod/app": "kong",
     "k8s-pod/pod-template-hash": "5b757c4bb8"
   },
   "logName": "projects/brokergenius-qa-env/logs/stderr",
   "sourceLocation": {
     "file": "queue.go",
     "line": "113"
   },
   "receiveTimestamp": "2020-02-03T09:48:10.615052191Z"
 },
 {
   "textPayload": "10.190.0.43 - - [03/Feb/2020:09:48:09 +0000] \"GET / HTTP/1.1\" 200 50 \"-\" \"GoogleHC/1.0\"\n",
   "insertId": "c9u31sosn0m6sanfs",
   "resource": {
     "type": "k8s_container",
     "labels": {
       "project_id": "brokergenius-qa-env",
       "cluster_name": "k8s-qa",
       "pod_name": "gateway-kong-5b757c4bb8-454mp",
       "container_name": "kong",
       "namespace_name": "gateway",
       "location": "us-east1-b"
     }
   },
   "timestamp": "2020-02-03T09:48:09.049491721Z",
   "severity": "INFO",
   "labels": {
     "k8s-pod/component": "app",
     "k8s-pod/release": "gateway",
     "k8s-pod/app": "kong",
     "k8s-pod/pod-template-hash": "5b757c4bb8"
   },
   "logName": "projects/brokergenius-qa-env/logs/stdout",
   "receiveTimestamp": "2020-02-03T09:48:10.353587697Z"
 },
 {
   "insertId": "umlngu6lf9ji0l935",
   "jsonPayload": {
     "source": "reflector.go:125",
     "message": "pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to list *v1.KongConsumer: Unauthorized",
     "pid": "1"
   },
   "resource": {
     "type": "k8s_container",
     "labels": {
       "container_name": "ingress-controller",
       "namespace_name": "gateway",
       "location": "us-east1-b",
       "project_id": "brokergenius-qa-env",
       "cluster_name": "k8s-qa",
       "pod_name": "gateway-kong-5b757c4bb8-krr95"
     }
   },
   "timestamp": "2020-02-03T09:48:09.043401Z",
   "severity": "ERROR",
   "labels": {
     "k8s-pod/release": "gateway",
     "k8s-pod/app": "kong",
     "k8s-pod/pod-template-hash": "5b757c4bb8",
     "k8s-pod/component": "app"
   },
   "logName": "projects/brokergenius-qa-env/logs/stderr",
   "sourceLocation": {
     "file": "reflector.go",
     "line": "125"
   },
   "receiveTimestamp": "2020-02-03T09:48:10.615052191Z"
 },
 {
   "insertId": "umlngu6lf9ji0l934",
   "jsonPayload": {
     "source": "reflector.go:125",
     "message": "pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to list *v1.KongCredential: Unauthorized",
     "pid": "1"
   },
   "resource": {
     "type": "k8s_container",
     "labels": {
       "location": "us-east1-b",
       "project_id": "brokergenius-qa-env",
       "cluster_name": "k8s-qa",
       "pod_name": "gateway-kong-5b757c4bb8-krr95",
       "container_name": "ingress-controller",
       "namespace_name": "gateway"
     }
   },
   "timestamp": "2020-02-03T09:48:09.042660Z",
   "severity": "ERROR",
   "labels": {
     "k8s-pod/pod-template-hash": "5b757c4bb8",
     "k8s-pod/component": "app",
     "k8s-pod/release": "gateway",
     "k8s-pod/app": "kong"
   },
   "logName": "projects/brokergenius-qa-env/logs/stderr",
   "sourceLocation": {
     "file": "reflector.go",
     "line": "125"
   },
   "receiveTimestamp": "2020-02-03T09:48:10.615052191Z"
 },
 {
   "insertId": "umlngu6lf9ji0l945",
   "jsonPayload": {
     "pid": "1",
     "source": "reflector.go:125",
     "message": "pkg/mod/k8s.io/client-go@v0.0.0-20190819141724-e14f31a72a77/tools/cache/reflector.go:98: Failed to list *v1.KongPlugin: Unauthorized"
   },
   "resource": {
     "type": "k8s_container",
     "labels": {
       "container_name": "ingress-controller",
       "namespace_name": "gateway",
       "location": "us-east1-b",
       "project_id": "brokergenius-qa-env",
       "pod_name": "gateway-kong-5b757c4bb8-hx44g",
       "cluster_name": "k8s-qa"
     }
   }
```

### Errors from kong
```json

   "textPayload": "2020/01/31 15:29:18 [warn] 35#0: *382897393 [lua] base.lua:562: queryDns(): [ringbalancer 859190] querying dns for marketdata-live-wolow-marketdata.marketdata failed: dns server error: 3 name error. Tried [\"(short)marketdata-live-wolow-marketdata.marketdata:(na) - cache-miss\",\"marketdata-live-wolow-marketdata.marketdata.gateway.svc.cluster.local:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.svc.cluster.local:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.cluster.local:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.c.brokergenius-qa-env.internal:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.google.internal:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata:33 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.gateway.svc.cluster.local:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.svc.cluster.local:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.cluster.local:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.c.brokergenius-qa-env.internal:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.google.internal:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata:1 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.gateway.svc.cluster.local:5 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.svc.cluster.local:5 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.cluster.local:5 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.c.brokergenius-qa-env.internal:5 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata.google.internal:5 - cache-miss/scheduled/querying/dns server error: 3 name error\",\"marketdata-live-wolow-marketdata.marketdata:5 - cache-miss/scheduled/querying/dns server error: 3 name error\"], context: ngx.timer, client: 127.0.0.1, server: 0.0.0.0:8444\n",
   "insertId": "c9u31sosn0m3cpgul",
   "resource": {
     "type": "k8s_container",
     "labels": {
       "location": "us-east1-b",
       "project_id": "brokergenius-qa-env",
       "pod_name": "gateway-kong-5b757c4bb8-454mp",
       "cluster_name": "k8s-qa",
       "container_name": "kong",
       "namespace_name": "gateway"
     }
   },
   "timestamp": "2020-01-31T15:29:18.798343150Z",
   "severity": "ERROR",
   "labels": {
     "k8s-pod/release": "gateway",
     "k8s-pod/app": "kong",
     "k8s-pod/pod-template-hash": "5b757c4bb8",
     "k8s-pod/component": "app"
   },
   "logName": "projects/brokergenius-qa-env/logs/stderr",
   "receiveTimestamp": "2020-01-31T15:29:20.529338880Z"
 },
```

### Workaround
For now, as a workaround, I `restarted` my pods and it `fixed` my issue but I would like to know what caused it and how can we avoid this in the future and help other people if they see a similar error. 

In General, moving to DB-less mode was a great idea and its quite stable.

### Steps To Reproduce

I don't know how to reproduce the error since I have no idea what might cause the issue. 

### Additional Details & Logs

- Kong version :  1.3
- Kong Ingress Controller :  0.6.0
- DB less Mode : Enabled
- Platform: GKE

**Comment by hbagdi** [comment]
This has been reported previously but we have been able to hunt down the problem. It is likely a bug in the version of `client-go` we use or the apiserver itself but it is hard to reproduce so we haven't been able to figure a solution out.
From the errors we see, this happens only for Custom Resources and never happens for core resources.

If it was an authentication issue, then it could have been due to tokens not being refreshed by `client-go` but that doesn't seem to be the case either.
